pkg/sql/alter_table_locality.go

```diff
@@ -28,5 +43,246 @@ func (p *planner) AlterTableLocality(
+			dbDesc,
+			n,
+		)
+	}
+
+	var newRegion *descpb.RegionName = nil
+	newRegionName := descpb.RegionName(n.n.Locality.TableRegion)
+
+	// If we haven't been provided a new region, we're altering to the PRIMARY REGION.
+	alterToPrimaryRegion := newRegionName == ""
+	// If we're altering to the primary region, leave the newRegion as nil.
+	if !alterToPrimaryRegion {
+		newRegion = &newRegionName
+	}
+
+	// In all cases if we get down here we need to update the Locality info in the table
+	// descriptor. It's possible here that the LocalityConfig is nil, if we're altering
+	// from a table that was not provided a table locality at creation. Allocate a
+	// LocalityConfig here to be used below. Note that all ALTERs which change the
+	// table locality will result in a LocalityConfig existing on the table (i.e. there
+	// currently is no way to remove the LocalityConfig from a table via ALTER).
+	if n.tableDesc.LocalityConfig == nil {
+		n.tableDesc.LocalityConfig = &descpb.TableDescriptor_LocalityConfig{}
+	}
+
+	// Setup the Locality and Region fields of the LocalityConfig. In cases where we're
+	// altering to the PRIMARY REGION, the Region will be set to nil so that SHOW CREATE
+	// TABLE will show "REGIONAL BY TABLE IN PRIMARY REGION".
+	n.tableDesc.LocalityConfig.Locality =
+		&descpb.TableDescriptor_LocalityConfig_RegionalByTable_{
+			RegionalByTable: &descpb.TableDescriptor_LocalityConfig_RegionalByTable{
+				Region: newRegion,
+			},
+		}
+
+	// Validate the new locality before updating the table descriptor.
+	tableName := tree.MakeTableName(tree.Name(dbDesc.Name), tree.Name(n.tableDesc.GetName()))
+	if err := tabledesc.ValidateTableLocalityConfig(
+		tableName.String(),
+		n.tableDesc.LocalityConfig,
+		dbDesc,
+	); err != nil {
+		return err
+	}
+
+	// Write out the table descriptor update.
+	if err := params.p.writeSchemaChange(
+		params.ctx,
+		n.tableDesc,
+		descpb.InvalidMutationID,
+		tree.AsStringWithFQNames(&n.n, params.Ann()),
+	); err != nil {
+		return err
+	}
+
+	// Update the table's zone configuration. If we're altering to the PRIMARY REGION,
+	// we don't want to set a zone configuration (in fact, we'll remove any existing zone
+	// configuration below).
+	if !alterToPrimaryRegion {
+		if err := params.p.applyZoneConfigFromTableLocalityConfig(
+			params.ctx,
+			tableName,
+			n.tableDesc.TableDesc(),
+			*dbDesc.RegionConfig,
+		); err != nil {
+			return err
+		}
+	} else {
+		// TODO(#multiregion): We should check to see if the zone configuration has been updated
+		// by the user. If it has, we need to warn, and only proceed if sql_safe_updates is disabled.
+
+		// Table is placed in the PRIMARY REGION. Remove the table's zone configuration so
+		// that it can be inherited from the database.
+		fullTableName := tree.MakeTableNameWithSchema(
+			tableName.CatalogName,
+			tableName.SchemaName,
+			tableName.ObjectName,
+		)
+		sql := fmt.Sprintf("ALTER TABLE %s CONFIGURE ZONE DISCARD", fullTableName.String())
```

Comment:
cc @ajstorm late night epiphanies just were had

just realised we probably don't want to discard here if the user modified other fields (e.g. range_min_bytes). can you add a TODO in a subsequent PR?